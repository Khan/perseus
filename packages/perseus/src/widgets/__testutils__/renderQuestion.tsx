import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
