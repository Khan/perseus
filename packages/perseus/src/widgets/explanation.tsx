/* eslint-disable react/sort-comp */
import {linterContextDefault} from "@khanacademy/perseus-linter";
import {StyleSheet, css} from "aphrodite";
import * as React from "react";
import _ from "underscore";

import * as Changeable from "../mixins/changeable";
import Renderer from "../renderer";
import * as styleConstants from "../styles/constants";
import mediaQueries from "../styles/media-queries";

import type {PerseusExplanationWidgetOptions} from "../perseus-types";
import type {PerseusScore, WidgetExports, WidgetProps} from "../types";

// TODO(FEI-5054): Figure out how to get global .d.ts files working with monorepos
type Empty = Record<never, never>;

type RenderProps = PerseusExplanationWidgetOptions; // transform = _.identity

type Rubric = PerseusExplanationWidgetOptions;

type UserInput = Empty;

type Props = WidgetProps<RenderProps, PerseusExplanationWidgetOptions>;

type DefaultProps = {
    showPrompt: Props["showPrompt"];
    hidePrompt: Props["hidePrompt"];
    explanation: Props["explanation"];
    widgets: Props["widgets"];
    linterContext: Props["linterContext"];
};

type State = {
    expanded: boolean;
};

class Explanation extends React.Component<Props, State> {
    static defaultProps: DefaultProps = {
        showPrompt: "Explain",
        hidePrompt: "Hide explanation",
        explanation: "explanation goes here\n\nmore explanation",
        widgets: {},
        linterContext: linterContextDefault,
    };

    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        return {
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        };
    }

    state: State = {
        expanded: false,
    };

    change: (arg1: any, arg2: any, arg3: any) => any = (...args) => {
        return Changeable.change.apply(this, args);
    };

    _onClick: () => void = () => {
        this.setState({
            expanded: !this.state.expanded,
        });
        this.props.trackInteraction();
    };

    render(): React.ReactNode {
        const {Link} = this.props.apiOptions.baseElements;
        const {isArticle, isMobile} = this.props.apiOptions;

        const linkAnchor = this.state.expanded
            ? this.props.hidePrompt
            : this.props.showPrompt;

        let linkContainer;

        // TODO(diedra): This isn't a valid href;
        // change this to a button that looks like a link.
        const href = "javascript:void(0)";
        const onClick = this._onClick;

        if (isMobile) {
            linkContainer = (
                <div className={css(styles.linkContainer)}>
                    <a
                        className={css(styles.mobileExplanationLink)}
                        href={href}
                        onClick={onClick}
                        role="button"
                        aria-expanded={this.state.expanded}
                    >
                        {linkAnchor}
                    </a>
                    {this.state.expanded && (
                        <svg className={css(styles.disclosureArrow)}>
                            <polygon
                                style={{fill: backgroundColor}}
                                points={
                                    `0,${arrowHeight} ` +
                                    `${arrowWidth},${arrowHeight} ` +
                                    `${arrowWidth / 2},0`
                                }
                            />
                        </svg>
                    )}
                </div>
            );
        } else {
            if (isArticle) {
                // NOTE: For articles, the baseElements `Link` may is either be
                // an anchor tag (if rendered in a modal) or a Wonder Blocks
                // link component in lesson page, see: `article-page.jsx`.
                linkContainer = (
                    <div
                        className={css(
                            styles.linkContainer,
                            styles.articleLink,
                        )}
                    >
                        <Link
                            className={css(styles.explanationLink)}
                            // HACK(michaelpolyak): WB Link doesn't passthrough
                            // class name.
                            style={styles.explanationLink}
                            href={href}
                            onClick={onClick}
                            role="button"
                            aria-expanded={this.state.expanded}
                        >
                            {`[${linkAnchor}]`}
                        </Link>
                    </div>
                );
            } else {
                // NOTE: For exercises, the baseElements `Link` is an
                // anchor tag, see: `perseus-api.jsx`.
                linkContainer = (
                    <div
                        className={css(
                            styles.linkContainer,
                            styles.exerciseLink,
                        )}
                    >
                        <Link
                            className={css(styles.explanationLink)}
                            href={href}
                            onClick={onClick}
                            role="button"
                            aria-expanded={this.state.expanded}
                        >
                            {`[${linkAnchor}]`}
                        </Link>
                    </div>
                );
            }
        }

        const expandedStyle = isMobile
            ? styles.contentExpandedMobile
            : styles.contentExpanded;

        return (
            <div className={css(styles.container)}>
                {linkContainer}
                {this.state.expanded && (
                    <div
                        className={css(
                            styles.content,
                            isMobile && styles.contentMobile,
                            this.state.expanded && expandedStyle,
                        )}
                        // eslint-disable-next-line react/no-string-refs
                        ref="content"
                    >
                        <Renderer
                            apiOptions={this.props.apiOptions}
                            content={this.props.explanation}
                            widgets={this.props.widgets}
                            linterContext={this.props.linterContext}
                        />
                    </div>
                )}
            </div>
        );
    }

    getUserInput: () => Empty = () => {
        return {};
    };

    simpleValidate: (arg1: Rubric) => PerseusScore = (rubric) => {
        return Explanation.validate(this.getUserInput(), rubric);
    };
}

const leftBorderSpacing = 23;
const verticalContentPadding = 10;

const arrowWidth = 30;
const arrowHeight = 14;
const backgroundColor = styleConstants.gray95;

const styles = StyleSheet.create({
    container: {
        display: "inline",
        position: "relative",
    },

    linkContainer: {
        display: "inline-block",
    },

    explanationLink: {
        fontStyle: "italic",
        color: "#007d96",
    },

    articleLink: {
        // Copied from .body-text in articles.less
        fontSize: 20,
        lineHeight: "30px",
    },

    exerciseLink: {
        // Copied from .legacy-typography in util.less
        fontSize: 14,
        lineHeight: "19.6px",
    },

    mobileExplanationLink: {
        color: styleConstants.kaGreen,
        borderBottom: `dashed 1px ${styleConstants.kaGreen}`,
        textDecoration: "none",

        // TODO(benkomalo): these should be pulled in from common typography
        // shared files so we have a single place where the type hierarchy is
        // defined; one off font sizes for individual components should be
        // avoided.
        [mediaQueries.xl]: {
            fontSize: 22,
            lineHeight: 1.4,
        },
        [mediaQueries.lgOrSmaller]: {
            fontSize: 20,
            lineHeight: 1.5,
        },
        [mediaQueries.smOrSmaller]: {
            fontSize: 18,
            lineHeight: 1.2,
        },
    },

    content: {
        position: "relative",
        transition: "margin-top 0.1s",
    },

    contentExpanded: {
        borderLeft: "5px solid #ccc",
        marginLeft: -leftBorderSpacing,
        paddingLeft: leftBorderSpacing,

        paddingBottom: verticalContentPadding,

        // Note: we still use arrow height as the vertical margin, even on
        // desktop when there is no arrow, but it's good enough.
        marginBottom: arrowHeight,
        marginTop: arrowHeight,
    },

    contentExpandedMobile: {
        boxSizing: "content-box",
        paddingTop: 32,
        paddingBottom: 32,
        marginTop: arrowHeight,
    },

    contentMobile: {
        background: backgroundColor,

        // TODO(benkomalo): this is to "full bleed" the background.
        // The actual content padding differs depending on the host
        // container, so this needs to be fixed eventually.
        marginLeft: styleConstants.negativePhoneMargin,
        marginRight: styleConstants.negativePhoneMargin,
        paddingLeft: styleConstants.phoneMargin,
        paddingRight: styleConstants.phoneMargin,
    },

    disclosureArrow: {
        // HACK - positioning at "bottom: 0", doesn't actually position it to
        // the real bottom, because the container is `inline-block`, and it
        // seems to position it to the baseline? We put in a generous
        // fudge factor to position it down to be flush with the content box
        // below it.
        bottom: -(arrowHeight + 5),
        height: arrowHeight,
        left: "50%",
        marginLeft: -(arrowWidth / 2),
        position: "absolute",
        width: arrowWidth,
    },
});

export default {
    name: "explanation",
    displayName: "Explanation",
    accessible: true,
    defaultAlignment: "inline",
    widget: Explanation,
    transform: _.identity,
    isLintable: true,
} as WidgetExports<typeof Explanation>;
