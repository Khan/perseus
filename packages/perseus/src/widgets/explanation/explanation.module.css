@layer reset, shared, legacy;

@layer legacy {

    /*
      Special styling is needed to fit the button in a block of text without throwing off the line spacing.
      While the button is not normally included in a block of text, it needs to be able to accommodate such a case.
    */
    .button-style-overrides {
        height: auto;
        line-height: inherit;
        margin-left: -2px; /* transfer space on the left side to the padding */
        margin-right: 2px; /* adds space between the right border and any text that follows */
        padding-left: 2px; /* adds space between the left border and the button text */
    }

    .button-style-overrides > span {
        font-size: 18px;
        line-height: inherit;
        text-align: left;
        /*
          The following property adjusts the large space between the button text and the caret icon.
          Since we are unable to adjust the styling of the icon (where the extra space exists),
          we are adjusting it on the text label by using a negative margin.
        */
        margin-right: -6px;
        white-space: normal;
    }

    .content {
        border-left: 0px solid #ccc;
        display: inline-grid;
        position: relative;
    }

    .content-collapsed {
        grid-template-columns: 0fr;
        grid-template-rows: 0fr;
        margin-bottom: 0;
        margin-top: 0;
        min-width: 0;
        padding-bottom: 0;
        visibility: hidden;
    }

    .content-expanded {
        border-left-width: 5px;
        grid-template-columns: 1fr;
        grid-template-rows: 1fr;
        margin-left: -23px;
        min-width: 100%;
        padding-left: 23px;
        padding-bottom: 10px;
        visibility: visible;
        /*
          Note: we still use arrow height as the vertical margin, even on
          desktop when there is no arrow, but it's good enough.
        */
        margin-bottom: 14px;
        margin-top: 14px;
    }

    .content-wrapper {
        overflow: hidden;
    }

    .transition-collapsed {
        transition: all 0.25s step-end, grid-template-rows 0.25s, margin-top 0.25s, margin-bottom 0.25s, padding-bottom 0.25s;
    }

    .transition-expanded {
        transition: grid-template-rows 0.5s, margin-top 0.5s, margin-bottom 0.5s, padding-bottom 0.5s;
    }

}
