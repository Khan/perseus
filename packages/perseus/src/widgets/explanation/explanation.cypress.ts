import renderQuestion from "../../../../../testing/render-question-with-cypress";
import {cypressTestDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import * as Perseus from "../../index";

import {ipsumExample} from "./explanation.testdata";

//  NOTE: The regression tests in this file use Cypress because they are intended to validate styling that is applied.
//        Since React Testing Library isn't applying the CSS to the elements,
//          we can't use Jest to verify that some keyboard interactions work properly.

describe("Explanation Widget", () => {
    const postContentLinkText = "the system J-25";
    const contentLinkOne = "synchronic distortion";
    const contentLinkTwo = "selective molecular polarization";

    beforeEach(() => {
        Dependencies.setDependencies(cypressTestDependencies);
        Perseus.init();
    });

    it("prevents interacting with actionable items within content when COLLAPSED (initial state)", () => {
        // Arrange
        renderQuestion(ipsumExample);
        cy.get("button[aria-expanded='false'][aria-controls]").focus();

        // Act - verify tab order (forwards)
        cy.focused().should("have.text", "Explanation"); // Verify we are on the widget's button.
        cy.realPress("Tab");
        cy.focused().should("have.text", postContentLinkText);

        // Act - verify tab order (backwards)
        cy.realPress(["Shift", "Tab"]);
        cy.focused().should("have.text", "Explanation");
    });

    it("allows interacting with actionable items within content when EXPANDED", () => {
        // NOTE: This test ensures that the CSS that controls keyboard access doesn't regress.
        //       It also ensures that any JavaScript event handling doesn't interfere with expected keyboard navigation.

        // Arrange
        renderQuestion(ipsumExample);
        cy.get("button[aria-expanded='false'][aria-controls]").focus();
        cy.focused().should("have.text", "Explanation"); // Verify we are on the widget's button.
        cy.realPress("Enter"); // Expand content

        // Act - verify tab order (forwards)
        cy.realPress("Tab");
        cy.focused().should("have.text", contentLinkOne);
        cy.realPress("Tab");
        cy.focused().should("have.text", contentLinkTwo);
        cy.realPress("Tab");
        cy.focused().should("have.text", postContentLinkText);

        // Act - verify tab order (backwards)
        cy.realPress(["Shift", "Tab"]);
        cy.focused().should("have.text", contentLinkTwo);
        cy.realPress(["Shift", "Tab"]);
        cy.focused().should("have.text", contentLinkOne);
        cy.realPress(["Shift", "Tab"]);
        cy.focused().should("have.text", "Hide"); // "Hide" is the new text in the widget's button.
    });

    it("prevents interacting with actionable items within content when COLLAPSED (after toggle)", () => {
        // NOTE: This test ensures that interaction with the widget's button doesn't regress.
        //       The test is similar to the first "COLLAPSED" test,
        //          but while that one tests the initial state of the widget,
        //          this one verifies that toggling doesn't introduce/remove anything that would interfere with
        //          expected keyboard navigation.

        // Arrange
        renderQuestion(ipsumExample);
        cy.get("button[aria-expanded='false'][aria-controls]").focus();
        cy.focused().should("have.text", "Explanation"); // Verify we are on the widget's button.
        cy.realPress("Enter"); // Expand content

        // Act - verify tab order (forwards)
        cy.realPress("Tab");
        cy.focused().should("have.text", contentLinkOne);

        // Act - verify tab order (backwards)
        cy.realPress(["Shift", "Tab"]);
        cy.realPress("Enter"); // Collapse content
        cy.waitUntil(() =>
            cy
                .focused()
                .siblings()
                .first()
                .should("have.attr", "aria-hidden", "true")
                .should("have.css", "visibility", "hidden"),
        );
        cy.realPress("Tab");
        cy.focused().should("have.text", postContentLinkText);
    });
});
