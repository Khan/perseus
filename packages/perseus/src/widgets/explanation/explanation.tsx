import {linterContextDefault} from "@khanacademy/perseus-linter";
import Button from "@khanacademy/wonder-blocks-button";
import {Id, View} from "@khanacademy/wonder-blocks-core";
import caretDown from "@phosphor-icons/core/regular/caret-down.svg";
import caretUp from "@phosphor-icons/core/regular/caret-up.svg";
import * as React from "react";
import _ from "underscore";

import {PerseusI18nContext} from "../../components/i18n-context";
import * as Changeable from "../../mixins/changeable";
import Renderer from "../../renderer";
import {getPromptJSON as _getPromptJSON} from "../../widget-ai-utils/explanation/explanation-ai-utils";
import scoreNoop from "../__shared__/score-noop";

import styles from "./explanation.module.css";

import type {PerseusExplanationWidgetOptions} from "../../perseus-types";
import type {Widget, WidgetExports, WidgetProps} from "../../types";
import type {ExplanationPromptJSON} from "../../widget-ai-utils/explanation/explanation-ai-utils";
import type {PerseusExplanationWidgetOptions} from "@khanacademy/perseus-core";

type RenderProps = PerseusExplanationWidgetOptions; // transform = _.identity

type Props = WidgetProps<RenderProps>;

type DefaultProps = {
    showPrompt: Props["showPrompt"];
    hidePrompt: Props["hidePrompt"];
    explanation: Props["explanation"];
    widgets: Props["widgets"];
    linterContext: Props["linterContext"];
};

type State = {
    expanded: boolean;
};

function mediaQueryIsMatched(mediaQuery: string): boolean {
    if (typeof window.matchMedia !== "function") {
        return false;
    }
    return window.matchMedia(mediaQuery).matches;
}

class Explanation extends React.Component<Props, State> implements Widget {
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    static defaultProps: DefaultProps = {
        showPrompt: "Explain",
        hidePrompt: "Hide explanation",
        explanation: "explanation goes here\n\nmore explanation",
        widgets: {},
        linterContext: linterContextDefault,
    };

    // this just helps with TS weak typing when a Widget
    // doesn't implement any Widget methods
    isWidget = true as const;

    state: State = {
        expanded: false,
    };
    _mounted: boolean = false;

    componentDidMount() {
        this._mounted = true;
    }

    componentWillUnmount() {
        this._mounted = false;
    }

    change: (arg1: any, arg2: any, arg3: any) => any = (...args) => {
        return Changeable.change.apply(this, args);
    };

    _onClick: () => void = () => {
        this.setState({
            expanded: !this.state.expanded,
        });
        this.props.trackInteraction();
    };

    getPromptJSON(): ExplanationPromptJSON {
        return _getPromptJSON(this.props);
    }

    render(): React.ReactNode {
        const promptText = this.state.expanded
            ? this.props.hidePrompt
            : this.props.showPrompt;

        const caretIcon = this.state.expanded ? caretUp : caretDown;

        const allowTransition =
            this._mounted &&
            mediaQueryIsMatched("(prefers-reduced-motion: no-preference)");

        const contentStyling = [
            styles.content,
            this.state.expanded
                ? styles.contentExpanded
                : styles.contentCollapsed,
            allowTransition &&
                (this.state.expanded
                    ? styles.transitionExpanded
                    : styles.transitionCollapsed),
        ];

        return (
            <Id>
                {(contentId) => (
                    <>
                        <Button
                            aria-expanded={this.state.expanded}
                            aria-controls={contentId}
                            endIcon={caretIcon}
                            kind="tertiary"
                            onClick={this._onClick}
                            size="small"
                            className={styles.buttonStyleOverrides}
                        >
                            {promptText}
                        </Button>

                        <View
                            id={contentId}
                            className={contentStyling.join(" ")}
                            aria-hidden={!this.state.expanded}
                            testId="content-container"
                        >
                            <View className={styles.contentWrapper}>
                                <Renderer
                                    apiOptions={this.props.apiOptions}
                                    content={this.props.explanation}
                                    widgets={this.props.widgets}
                                    linterContext={this.props.linterContext}
                                    strings={this.context.strings}
                                />
                            </View>
                        </View>
                    </>
                )}
            </Id>
        );
    }
}

export default {
    name: "explanation",
    displayName: "Explanation",
    accessible: true,
    defaultAlignment: "inline",
    widget: Explanation,
    transform: _.identity,
    isLintable: true,
    // TODO: things that aren't interactive shouldn't need scoring functions
    scorer: () => scoreNoop(),
} satisfies WidgetExports<typeof Explanation>;
