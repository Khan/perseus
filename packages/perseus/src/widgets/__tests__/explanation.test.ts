import {screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

// TODO(FEI-3857): Include in jest setup so that we don't need to import it everywhere
import "@testing-library/jest-dom/extend-expect";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {question1} from "../__testdata__/explanation.testdata";
import ExplanationWidgetExports from "../explanation";

import {renderQuestion} from "./renderQuestion";

import type {PerseusExplanationWidgetOptions} from "../../perseus-types";

describe("Explanation", function () {
    beforeEach(() => {
        // We mock out `console.error` because the explanation widget adds
        // `javascript:void(0);` to the `onClick` handler which triggers an
        // error in our test env.
        jest.spyOn(console, "error").mockImplementation(() => {});

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange and Act
        const {container} = renderQuestion(question1);

        // Assert
        expect(container).toMatchSnapshot("initial render");
    });

    it("should snapshot when expanded", () => {
        // Arrange
        const {container} = renderQuestion(question1);

        // Act
        userEvent.click(screen.getByRole("button"));

        // Assert
        expect(container).toMatchSnapshot("expanded");
    });

    it("should snapshot for mobile", () => {
        // Arrange and Act
        const {container} = renderQuestion(question1, {
            isMobile: true,
        });

        // Assert
        expect(container).toMatchSnapshot("initial render");
    });

    it("should snapshot for mobile when expanded", () => {
        // Arrange
        const {container} = renderQuestion(question1, {
            isMobile: true,
        });

        // Act
        userEvent.click(screen.getByRole("button"));

        // Assert
        expect(container).toMatchSnapshot("expanded");
    });

    it("should snapshot for article", () => {
        // Arrange and Act
        const {container} = renderQuestion(question1, {
            isArticle: true,
        });

        // Assert
        expect(container).toMatchSnapshot("initial render");
    });

    it("should snapshot for article when expanded", () => {
        // Arrange
        const {container} = renderQuestion(question1, {
            isArticle: true,
        });

        // Act
        userEvent.click(screen.getByRole("button"));

        // Assert
        expect(container).toMatchSnapshot("expanded");
    });

    it("should snapshot for article+mobile", () => {
        // Arrange and Act
        const {container} = renderQuestion(question1, {
            isMobile: true,
            isArticle: true,
        });

        // Assert
        expect(container).toMatchSnapshot("initial render");
    });

    it("should snapshot for article+mobile when expanded", () => {
        // Arrange
        const {container} = renderQuestion(question1, {
            isMobile: true,
            isArticle: true,
        });

        // Act
        userEvent.click(screen.getByRole("button"));

        // Assert
        expect(container).toMatchSnapshot("expanded");
    });

    it("can be expanded", function () {
        // Arrange
        renderQuestion(question1);

        // Act
        const expandLink = screen.getByRole("button", {expanded: false});
        userEvent.click(expandLink);

        // Assert
        // get asserts if it doesn't find a single matching element
        expect(screen.getByText("This is an explanation")).toBeVisible();
    });

    it("can be collapsed", function () {
        // Arrange
        renderQuestion(question1);

        // Act
        const expandLink = screen.getByRole("button", {expanded: false});
        userEvent.click(expandLink); // expand and then
        const collapseLink = screen.getByRole("button", {expanded: true});
        userEvent.click(collapseLink); // collapse

        // Assert
        expect(screen.queryByText("This is an explanation")).toBeNull();
    });

    it("should return an empty object for getUserInput()", () => {
        // Arrange
        const {renderer} = renderQuestion(question1);

        // Act
        const userInput = renderer.getUserInput();

        // Assert
        expect(userInput).toMatchInlineSnapshot(`
            [
              {},
            ]
        `);
    });

    it("should return a zero score for simpleValidate()", () => {
        // Arrange
        const {renderer} = renderQuestion(question1);

        // Act
        const score = renderer.scoreWidgets();

        // Assert
        expect(score).toMatchInlineSnapshot(`
            {
              "explanation 1": {
                "earned": 0,
                "message": null,
                "total": 0,
                "type": "points",
              },
            }
        `);
    });

    describe("static validate", () => {
        it("should always return 0 points", () => {
            const result = ExplanationWidgetExports.widget.validate(
                {},
                question1.widgets["explanation 1"]
                    .options as PerseusExplanationWidgetOptions,
            );

            expect(result).toMatchInlineSnapshot(`
                {
                  "earned": 0,
                  "message": null,
                  "total": 0,
                  "type": "points",
                }
            `);
        });
    });
});
