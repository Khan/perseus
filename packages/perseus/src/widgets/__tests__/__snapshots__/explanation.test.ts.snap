// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explanation should snapshot for article when expanded: expanded 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <div
            class="container_167wsvl"
          >
            <div
              class="linkContainer_36rlri-o_O-articleLink_17wlykh"
            >
              <a
                aria-expanded="true"
                class="explanationLink_7n4oib"
                href="javascript:void(0)"
                role="button"
              >
                [Hide explanation!]
              </a>
            </div>
            <div
              class="content_z5b02x-o_O-contentExpanded_1c06cap"
            >
              <div
                class="perseus-renderer perseus-renderer-responsive"
              >
                <div
                  class="paragraph"
                  data-perseus-paragraph-index="0"
                >
                  <div
                    class="paragraph"
                  >
                    This is an explanation
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;

exports[`Explanation should snapshot for article+mobile when expanded: expanded 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <div
            class="container_167wsvl"
          >
            <div
              class="linkContainer_36rlri"
            >
              <a
                aria-expanded="true"
                class="mobileExplanationLink_his27d"
                href="javascript:void(0)"
                role="button"
              >
                Hide explanation!
              </a>
              <svg
                class="disclosureArrow_12nx0hv"
              >
                <polygon
                  points="0,14 30,14 15,0"
                  style="fill: #F0F1F2;"
                />
              </svg>
            </div>
            <div
              class="content_z5b02x-o_O-contentMobile_1wvx5oh-o_O-contentExpandedMobile_knze7z"
            >
              <div
                class="perseus-renderer perseus-renderer-responsive"
              >
                <div
                  class="paragraph"
                  data-perseus-paragraph-index="0"
                >
                  <div
                    class="paragraph"
                  >
                    This is an explanation
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;

exports[`Explanation should snapshot for article+mobile: initial render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <div
            class="container_167wsvl"
          >
            <div
              class="linkContainer_36rlri"
            >
              <a
                aria-expanded="false"
                class="mobileExplanationLink_his27d"
                href="javascript:void(0)"
                role="button"
              >
                Explanation
              </a>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;

exports[`Explanation should snapshot for article: initial render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <div
            class="container_167wsvl"
          >
            <div
              class="linkContainer_36rlri-o_O-articleLink_17wlykh"
            >
              <a
                aria-expanded="false"
                class="explanationLink_7n4oib"
                href="javascript:void(0)"
                role="button"
              >
                [Explanation]
              </a>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;

exports[`Explanation should snapshot for mobile when expanded: expanded 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <div
            class="container_167wsvl"
          >
            <div
              class="linkContainer_36rlri"
            >
              <a
                aria-expanded="true"
                class="mobileExplanationLink_his27d"
                href="javascript:void(0)"
                role="button"
              >
                Hide explanation!
              </a>
              <svg
                class="disclosureArrow_12nx0hv"
              >
                <polygon
                  points="0,14 30,14 15,0"
                  style="fill: #F0F1F2;"
                />
              </svg>
            </div>
            <div
              class="content_z5b02x-o_O-contentMobile_1wvx5oh-o_O-contentExpandedMobile_knze7z"
            >
              <div
                class="perseus-renderer perseus-renderer-responsive"
              >
                <div
                  class="paragraph"
                  data-perseus-paragraph-index="0"
                >
                  <div
                    class="paragraph"
                  >
                    This is an explanation
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;

exports[`Explanation should snapshot for mobile: initial render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <div
            class="container_167wsvl"
          >
            <div
              class="linkContainer_36rlri"
            >
              <a
                aria-expanded="false"
                class="mobileExplanationLink_his27d"
                href="javascript:void(0)"
                role="button"
              >
                Explanation
              </a>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;

exports[`Explanation should snapshot when expanded: expanded 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <div
            class="container_167wsvl"
          >
            <div
              class="linkContainer_36rlri-o_O-exerciseLink_574wt1"
            >
              <a
                aria-expanded="true"
                class="explanationLink_7n4oib"
                href="javascript:void(0)"
                role="button"
              >
                [Hide explanation!]
              </a>
            </div>
            <div
              class="content_z5b02x-o_O-contentExpanded_1c06cap"
            >
              <div
                class="perseus-renderer perseus-renderer-responsive"
              >
                <div
                  class="paragraph"
                  data-perseus-paragraph-index="0"
                >
                  <div
                    class="paragraph"
                  >
                    This is an explanation
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;

exports[`Explanation should snapshot: initial render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Here's the explanation

        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <div
            class="container_167wsvl"
          >
            <div
              class="linkContainer_36rlri-o_O-exerciseLink_574wt1"
            >
              <a
                aria-expanded="false"
                class="explanationLink_7n4oib"
                href="javascript:void(0)"
                role="button"
              >
                [Explanation]
              </a>
            </div>
          </div>
        </div>
        
Did you get that?
      </div>
    </div>
  </div>
</div>
`;
